src/libponyrt/actor/actor.c

```diff
@@ -443,10 +467,55 @@ PONY_API void pony_sendv_single(pony_ctx_t* ctx, pony_actor_t* to,
     DTRACE2(ACTOR_MSG_SEND, (uintptr_t)ctx->scheduler, last->id);
   }
 
+  maybe_mute(ctx, to, first, last);
+
   if(ponyint_messageq_push_single(&to->q, first, last))
   {
-    if(!has_flag(to, FLAG_UNSCHEDULED))
+    if(!has_flag(to, FLAG_UNSCHEDULED) && (to->muted == 0)) {
+      // if the receiving actor is currently unscheduled AND it's not
+      // muted, schedule it.
       ponyint_sched_add(ctx, to);
+    }
+  }
+}
+
+void maybe_mute(pony_ctx_t* ctx, pony_actor_t* to, pony_msg_t* first,
+  pony_msg_t* last)
+{
+  if(ctx->current != NULL)
+  {
+    // only mute a sender IF:
+    // 1. the receiver is overloaded or the receiver is muted
+    // AND
+    // 2. the sender isn't overloaded
+    // AND
+    // 3. we are sending to another actor (as compared to sending to self)
+    if((has_flag(to, FLAG_OVERLOADED) || (to->muted > 0)) &&
+      !has_flag(ctx->current, FLAG_OVERLOADED) &&
+      ctx->current != to)
+    {
+      if (first == last)
+      {
+        if(first->id <= ACTORMSG_APPLICATION_START)
+        {
+          ponyint_sched_mute(ctx, ctx->current, to);
+          return;
+        }
+      }
+
+      pony_msg_t* m = first;
+
+      while(m != last)
```

Comment:
Iterating the message chain here is going to be expensive, so I think it would be nice to refactor in order to remove the need for the loop.

In the current state of things, a message chain cannot contain ORCA messages, so a possible alternative would be to take the message chain length as a parameter. If we want to future-proof the function now, the parameter could be the number of application messages in the chain instead.

Reply:
If we use that assumption about message chains not containing ORCA messages, we should likely try to add a `pony_assert` somewhere to ensure that.

Reply:
That's true. I think the best place for that assertion would be in `pony_chain`.

Reply:
@Praetonus if we had the number of application messages in the chain, that would become much easier. if there is more than 0, we do our check. if there are none, then we don't need our check. i was going to bring that up.

Reply:
Discussed this with @SeanTAllen, I'm going to implement that change (counting the number of application messages in the chain) since it also requires a change to one of the optimisation passes, and I'll submit the patch in this PR.

Reply:
Praetonus' patch was applied.